src/dtls13.c

```diff
@@ -0,0 +1,996 @@
+
+    ret = wc_AesSetKey(&aes, key, ssl->specs.key_size, NULL, AES_ENCRYPTION);
+    if (ret != 0)
+        goto exit;
+
+    ret = wc_AesEncryptDirect(&aes, mask, ciphertext);
+    if (ret != 0)
+        goto exit;
+
+ exit:
+    wc_AesFree(&aes);
+    return ret;
+}
+
+static int dtls13_encrypt_decrypt_record_number(WOLFSSL *ssl,
+                                                byte *seq,
+                                                int seq_length,
+                                                byte *ciphertext,
+                                                enum rn_direction dir)
+
+{
+    byte mask[RN_MASK_SIZE];
+    int ret, i;
+    word8 side;
+
+    if (dir == PROTECT)
+        side = ssl->options.side;
+    else
+        side = (ssl->options.side == WOLFSSL_SERVER_END) ?
+            WOLFSSL_CLIENT_END : WOLFSSL_SERVER_END;
+
+    ret = dtls13_get_rn_mask(ssl, ciphertext, mask, side);
+    if (ret != 0)
+        return ret;
+
+    for (i = 0; i < seq_length; ++i)
+        seq[i] ^= mask[i];
+
+    return 0;
+}
+
+/* this is a clone of DtlsMsgDrain */
+static int dtls13_process_buffered_messages(WOLFSSL *ssl)
+{
+    DtlsMsg *msg = ssl->dtls_rx_msg_list;
+    word32 idx = 0;
+    int ret = 0;
+
+    WOLFSSL_ENTER("dtls13_process_buffered_messages()");
+    while (msg != NULL) {
+
+        /* message not in order */
+        if (ssl->keys.dtls_expected_peer_handshake_number != msg->seq)
+            break;
+
+        /* message not complete */
+        if (msg->fragSz != msg->sz)
+            break;
+
+        ret = DoTls13HandShakeMsgType(ssl, msg->msg, &idx, msg->type,
+                                      msg->sz, msg->sz);
+        if (ret != 0)
+            break;
+
+        /* TODO: update the retransmission logic with new message processed */
+
+        /* FIXME: move in a better place. handle overflow? */
+        ssl->keys.dtls_expected_peer_handshake_number++;
+
+        ssl->dtls_rx_msg_list = msg->next;
+        DtlsMsgDelete(msg, ssl->heap);
+        msg = ssl->dtls_rx_msg_list;
+        ssl->dtls_rx_msg_list_sz--;
+    }
+
+    WOLFSSL_LEAVE("dtls13_process_buffered_messages()", ret);
+
+    return ret;
+}
+
+static int dtls13_next_message_complete(WOLFSSL *ssl) {
+    return ssl->dtls_rx_msg_list != NULL &&
+        ssl->dtls_rx_msg_list->fragSz >= ssl->dtls_rx_msg_list->sz;
+}
+
+static inline int frag_is_in_output_buffer(WOLFSSL *ssl, const byte *frag)
+{
+    const byte *output_buffer = ssl->buffers.outputBuffer.buffer;
+    word16 output_buffer_size = ssl->buffers.outputBuffer.bufferSize;
+
+    return frag >= output_buffer && frag < output_buffer + output_buffer_size;
+}
+
+static int dtls13_send_frag_from_buffer(WOLFSSL *ssl,
+                                        byte *output, word16 length)
+{
+    byte *buf;
+    int ret;
+
+    ret = CheckAvailableSize(ssl, length);
+    if (ret != 0)
+        return ret;
+
+    buf = ssl->buffers.outputBuffer.buffer +
+        ssl->buffers.outputBuffer.length;
```

Comment:
Would it make sense to add a `frag_is_in_output_buffer` check here as well? To make sure that in the future you don't accidentally pass in `outputBuffer` in `output`.

Reply:
I think this will be invoked *only* after
```
    if (!frag_is_in_output_buffer(ssl, output))
        return dtls13_send_frag_from_buffer(ssl, output, length);
 ```
 this check. So it should be fine.